### ISSUE TITLE
catalog: add the Application kind

### ISSUE THREAD

**Issue by freben** [backstage/backstage#4873]
## Feature Suggestion

Internally at Spotify, the system model has evolved to also incorporate an entity kind `Application`.

![image](https://user-images.githubusercontent.com/3097461/110331627-a4134580-801f-11eb-8eb2-9e96bdc4bdce.png)

The summary blurb is as follows:

> An application is a bundle of components that is packaged, released and operated as one unit. Compared to a system, an application might host components that belong to different systems or are owned by different teams in addition to its own components. Typical examples would be a mobile application or the Backstage frontend.

Feature Request: Add this kind to the core model.

## Possible Implementation

This kind is not in widespread use yet. It is effectively being settled, and is hot off the presses. But it has sprung out of a real need to express these entities as distinct things.

We can illustrate the usage of this kind with a mobile client monorepo. The repo contains a large number of libraries and shared application features in their respective directories, and a smaller number of end user applications also in their respective directories. The latter require the former, and link them together to become the final applications.

```
├ features/
│  ├ feature1/
│  │  └ catalog-info.yaml  # F1 - Component of type app-feature
│  └ feature2/
│     └ catalog-info.yaml  # F2 - Component of type app-feature
│
├ lib/
│  └ library1/
│     └ catalog-info.yaml  # L3 - Component of type library
│
├ application1/
│  └ catalog-info.yaml     # A1 - Application of type mobile-app
│
├ application2/
│  └ catalog-info.yaml     # A2 - Application of type mobile-app
│
└ catalog-info.yaml        # R -  Component of type repository
```

There will be [relations](https://backstage.io/docs/features/software-catalog/well-known-relations) between features/libraries, applications, and the repository, as follows. 

- `F1 partOf R` (and conversely, `R hasPart F1`)
- `A partOf R` (and conversely, `R hasPart A`)
- `A hasPart F1` (and conversely, `F1 partOf A`)

So both the features and the application declare that they are part of the larger monorepo. This is mostly a structural fact - and it could actually be computed automatically by processors if one so desires. But the application explicitly states that it is composed of various features, and this is likely something that the application states in its spec, rather than the other way around.

## Risks

This is a "young" kind. Is it too soon to introduce?

There is a risk of confusion between using `Application` vs. for example `Component` with type `website`. When is a thing a component and when is it an application? Does this change how we feel about modelling services too, is a microservice an Application? Applications can't have APIs - or can they?

Are the relations the right ones? Applications could `dependsOn` instead of `hasPart` feature components, for example. However, it feels to me that the `dependsOn` relation implies a much more "external" dependency - something that probably is managed elsewhere that you call out to - rather than something that you actually are composed of.

## Alternative Solutions

Use the kind `Component`, with a type `application` or similar. This hinders the ability to subtype. It also leads to some level of unclarity what to do with the relations - we may then want to invent completely new relations for an application to its constituent features and libraries.

## Context

Users are already starting to model their monorepos and complex applications using the catalog, and there is certainly a need for _some_ guidance on how to express large, complex applications and their constituent parts.

**Comment by dtuite** [comment]
I believe it's healthy to push back on new additions to the model in the name of constraining complexity so I'm going to play devils advocate. 😄 

Could an application be modeled as a `type` of `System`?

> A system, in this sense, is a collection of resources and components that exposes one or several public APIs. The main benefit of modelling a system is that it hides its resources and private APIs between the components for any consumers.
[source](https://backstage.io/docs/features/software-catalog/system-model)

I see you have covered this already to an extent,

> Compared to a system, an application might host components that belong to different systems or are owned by different teams in addition to its own components. 

But couldn't a component b in two different systems at once to handle this use case?

**Comment by Fox32** [comment]
We have something in our organization that perfectly matches this (@petergrau). It's a mobile app in a mono repo and is build by multiple teams and combines a lot of systems and domains. The app itself is developed as a root application that consumes a lot of packages developed by different teams and that are part of different systems. Right now planned to model it using a component for the root application (type website?) and a library component for each package. The idea is to use dependsOn for modeling the relationship, as the root app has dependencies on the packages.

@freben You mention this approach in alternative solutions, can you give some more insights why this is a bad idea? I think I might got what the issue is, but I'm not sure. Component of type application, but you would like distinguish between android and iOS?

**Comment by petergrau** [comment]
Sounds a lot like the setup we have for our `app(s)` where the `app(s)` and their `libs` are part of one monorepo (nx) and deployed as one unit in different variations. The `libs` itself are all part of the monorepo but code ownership belongs to different teams. 

Form a first read I guess the question: What is a website and what is an application is an interesting one. However with the current "website" our impression was that it does not really describe what our `app(s)` are.

**Comment by andrewthauer** [comment]
We also have a use case this proposal describes quite closely. We use `micro-frontends` which are broken down as features of a larger `frontend` web application. In this case our top level `web` app could be an `Application` that consists of shared libraries, feature components. Many of the "feature" components we have interact with domain specific services which likely could map to domain entities (although we haven't modelled those out yet).

However, I also think we could likely model this using `subComponentOf` as well probably. But we would somehow need to extend the usage of `spec.type` to give a more appropriate context. Perhaps describing some of this can also come down to having additional filters and fields within the catalog and entity pages though?

**Comment by freben** [comment]
Yeah I think part of the point here with Application is that it can have its own schema, its own `spec` - which could give it the unique ability to point out its constituent app features etc. in a way that gets translated to this particular relation pair.

**Comment by freben** [comment]
> I believe it's healthy to push back on new additions to the model in the name of constraining complexity so I'm going to play devils advocate. 😄
> 
> Could an application be modeled as a `type` of `System`?

Hm. What type would a "regular" system be?

In the internal docs for the model, the architects have collected Domain and System under their own heading as "grouping concepts" so to speak, rather than software modeling concepts, so that they are to be regarded as a different dimension than the others.

> > Compared to a system, an application might host components that belong to different systems or are owned by different teams in addition to its own components.
> 
> But couldn't a component b in two different systems at once to handle this use case?

How do you mean with this last sentence?

**Comment by freben** [comment]
> Right now planned to model it using a component for the root application (type website?) and a library component for each package. The idea is to use dependsOn for modeling the relationship, as the root app has dependencies on the packages.
> 
> @freben You mention this approach in alternative solutions, can you give some more insights why this is a bad idea? I think I might got what the issue is, but I'm not sure.

We have gradually landed in a place where the `relationType + target.kind` pair of a relation is significant. If we use `Component dependsOn Component` to mean "I rely on this (external?) thing to function", then that will clash with the intent here, "I am a composite deliverable that's composed of these things". Does that make sense? Hence my comment in the Alternative Solutions section, that that probably implies a new relation type being needed.

Note also what this implies for the naming and documentation of the `Component#spec.subcomponentOf` field that generates the relation in question. Now it suddenly sounds awkward and one gets tempted to rename it and change its docs. Right? That tells me that the intents probably are very different.

This is tricky :)

> Component of type application, but you would like distinguish between android and iOS?

Yeah that's also an open question; what will a typical set of well known `Application#spec.type` values look like?

**Comment by freben** [comment]
> Form a first read I guess the question: What is a website and what is an application is an interesting one. However with the current "website" our impression was that it does not really describe what our `app(s)` are.

Yeah. Is a small single-purpose website different from for example a massive Backstage conglomerate, or the Spotify web player that's extremely client-like in all ways interpreting the word?

Does this mean that as a website grows, one may be tempted to change its kind? That would be very unfortunate.

I am on the fence about websites specifically as you probably can tell. It feels to me like we have to take a stance upfront - are all "runtime deliverable" things Applications, including services, websites and mobile apps - or not? I.e. roots of actual execution. Or is it not about that at all but rather something else?

**Comment by steelrooter** [comment]
Thanks for bringing this up @freben as this is really important to all mobile-first orgs like ours. We are also trying to model mobile apps and our model aligns well with what you have described.

A few questions:

> > Component of type application, but you would like distinguish between android and iOS?
> 
> Yeah that's also an open question; what will a typical set of well known `Application#spec.type` values look like?

Related to the iOS/Android `type` discussion, how should we model Flutter apps? Some apps that might be 95% Flutter, and then some might be 20% Flutter, so would we treat them differently?. Would Flutter app developers like to see related info in one place?

May be not directly related to this discussion, but what differentiates an Application Component from a Feature Component?

**Comment by mbruggmann** [comment]
Not sure I'm qualified to have an opinion on this, I haven't worked on these large monolithic applications much! But one thing I would add: We wouldn't want to introduce a new kind or component type just for modelling's sake, right? 

What we have at Spotify is that certain aspects like our release process, quality metrics, or bug tracking are attached to the application and not a specific feature or system. So we need some entity in Backstage to attach the plugins for these aspects to. Not sure if that informs the component vs system vs application way of modelling it, I don't have a strong opinion on that.

Then there is the question of if we should standardize this in the Backstage OSS model. I think that makes sense if we want to share plugins built for applications between organizations, or ship it with Backstage out of the box. Although I still feel like we should keep the basic model as simple as possible. Meaning, the core Backstage experience (like the catalog) should probably be usable without forcing every integrator or customer to adopt the application model. That's where the internal difference between the core entities and the "grouping concepts" comes from: the grouping concepts should be optional, and might not be implemented everywhere. It should be fine to not use applications on web for whatever reason for example.

**Comment by dtuite** [comment]
> Although I still feel like we should keep the basic model as simple as possible. Meaning, the core Backstage experience (like the catalog) should probably be usable without forcing every integrator or customer to adopt the application model. That's where the internal difference between the core entities and the "grouping concepts" comes from: the grouping concepts should be optional, and might not be implemented everywhere. It should be fine to not use applications on web for whatever reason for example.

This is a much better expression of what I was trying to say. I worry there's a real danger of overwhelming new Backstage users with options when they are trying to get started. 

It's nice for new users to be able to get started quickly with a small set of default kinds (User, Group, Component, Resource and System). Later, when they are more comfortable with Backstage, they can learn about types to more accurately model their architecture by tweaking the existing kinds. Introducing new kinds steepens the learning curve and discourages new users IMHO.

With that said, there is positive feedback from Backstage users above me in this comment thread so 🤷‍♂️ .

**Comment by m5m1th** [comment]
Looking at your description and comparing it to your [ecosystem model](https://backstage.io/docs/features/software-catalog/system-model#ecosystem-modeling), are you trying to capture the UI analogue to the API entity?

**Comment by andrewthauer** [comment]
> The idea is to use dependsOn for modeling the relationship, as the root app has dependencies on the packages.

@Fox32 - How does one use the `dependsOn` relation? I see it in the code and documented in a couple of spots, but it feels like it's an incomplete implementation right now.

**Comment by andrewthauer** [comment]
> You can't use it right now. You could emit it in a processor, but there is no way to use it in YAML and no way to display it. Till now it's just a plan to do it. We're thought about crawling our npm registry for the dependencies instead of doing it manually in YAML.

Thanks! We've thought about something similar in regards crawling our private registry. This would probably require annotating the related component entity with the package/dependency I suspect though.

I can also see the use of `dependsOn` for service to service dependencies that may not have API specs defined.

**Comment by andrewthauer** [comment]
I'm also sharing the sentiment of @dtuite in regards to not overloading users with too many concepts. Up until now it's been reasonable for teams to understand the concepts around components and APIs. I've recently been promoting the ideas of eco-system modelling and while folks see the benefit, it's not as easy to grok how to apply this in practice without a lot of background knowledge in the backstage catalog model. Especially with the sometimes subjectiveness of modelling choices available. Don't get me wrong I think there is huge power in what's in place, but educating, standardizing even what's currently in place will not be trivial within each org I suspect.

**Comment by steelrooter** [comment]
@freben Any update on this? Our org would benefit from having this kind added.

**Comment by freben** [comment]
Hey everybody, and sorry for the delay in getting back.

The general consensus seems to be that this is indeed a very useful (sometimes, almost necessary) construct, but it also adds significant confusion for orgs whose catalog structure just does not warrant the additional complexity.

Therefore, the vote is on _not_ adding this kind to the core kinds or exposing it to everybody out of the box. But at the same time, it would be an excellent tool to be able to add easily. So we need to figure out a way of making this usable but not adding to the confusion.

The plugin structure of the backend (and the catalog) is still fairly young. We don't have a very straightforward way of adding all of the aspects of a new kind as a single nice packaged deliverable. But I think that that might be a good next step: to figure out how to make the Application kind documented, and pluggable, with a minimum of fuss for those who need it.

Maybe we should start by making a new package `@backstage/plugin-catalog-kind-application-{react,common,node}` or similar. The type and json-schema could be in the common package, the node package could house the required processor(s) etc, the react package may contain some type guards and similar for the frontend to use.

**Comment by freben** [comment]
Actually removing the stale flag once more - I still want to invent some form of easy bundling that makes it simple to add everything needed for a new kind.

**Comment by freben** [comment]
I think we have a plan for this, in a similar vein tohttps://github.com/backstage/backstage/pull/7469

**Comment by freben** [comment]
Very early thoughts about bundling things in the catalog https://github.com/backstage/backstage/pull/7636
